lib/timeline/timeline_protocol.dart

```diff
@@ -165,214 +112,205 @@ class TimelineThread implements Comparable<TimelineThread> {
-
-  void format(StringBuffer buf, String indent) {
-    buf.writeln('$indent$name [${startMicros}u]');
-    for (TimelineThreadEvent child in children) {
-      child.format(buf, '  $indent');
+  TimelineEventType _getTimelineEventType(TraceEvent event) {
+    // The event will only be of type 'cpu' or 'gpu' because we do not process
+    // events of any other type.
+    if (event.isCpuEvent) {
+      return TimelineEventType.cpu;
+    } else {
+      return TimelineEventType.gpu;
     }
   }
-
-  @override
-  String toString() => '$name, start=$startMicros duration=$durationMicros';
 }
 
 class TimelineFrame {
-  TimelineFrame({this.renderStart, this.rasterizeStart});
+  TimelineFrame(this.id, this.startTime);
 
-  int renderStart;
-  int renderDuration;
+  final int id;
 
-  int rasterizeStart;
-  int rasterizeDuration;
+  final List<TimelineEvent> cpuEvents = [];
+  final List<TimelineEvent> gpuEvents = [];
 
-  int get startMicros => renderStart ?? rasterizeStart;
+  /// Frame start time in micros.
+  final int startTime;
 
-  int get endMicros {
-    if (rasterizeStart != null) {
-      return rasterizeStart + rasterizeDuration;
-    } else {
-      return renderStart + renderDuration;
-    }
-  }
+  /// Frame end time in micros.
+  int endTime;
 
-  void setRenderStart(int micros) {
-    renderStart = micros;
-  }
+  // TODO(kenzie): investigate why we are getting negative duration values
+  // TODO(kenzie): once correct frame data is available from the engine, verify
+  //  we have non-zero values for cpu/gpu durations when expected.
 
-  void setRenderEnd(int micros) {
-    renderDuration = micros - renderStart;
-  }
+  /// Duration the frame took to render in micros.
+  int get duration => endTime - startTime;
 
-  void setRasterizeStart(int micros) {
-    rasterizeStart = micros;
+  // Timing info for CPU portion of the frame.
+  int get cpuStartTime => cpuEvents.isNotEmpty ? cpuEvents.first.startTime : 0;
+  int get cpuDuration {
+    return cpuEvents.isNotEmpty
+        ? cpuEvents.last.startTime + cpuEvents.last.duration - cpuStartTime
+        : 0;
   }
 
-  void setRasterizeEnd(int micros) {
-    if (rasterizeStart != null) {
-      rasterizeDuration = micros - rasterizeStart;
-    }
+  int get cpuEndTime => cpuStartTime + cpuDuration;
+
+  // Timing info for GPU portion of the frame.
+  int get gpuStartTime => gpuEvents.isNotEmpty ? gpuEvents.first.startTime : 0;
+  int get gpuDuration {
+    return gpuEvents.isNotEmpty
+        ? gpuEvents.last.startTime + gpuEvents.last.duration - gpuStartTime
+        : 0;
   }
 
-  bool get isComplete => renderDuration != null && rasterizeDuration != null;
+  int get gpuEndTime => gpuStartTime + gpuDuration;
 
-  String get renderAsMs {
-    return '${(renderDuration / 1000.0).toStringAsFixed(1)}ms';
+  bool get isComplete => cpuDuration != null && gpuDuration != null;
+
+  String get cpuAsMs {
+    return '${(cpuDuration / 1000.0).toStringAsFixed(1)}ms';
```

Comment:
nit: add a helper method to display a duration as text.

Reply:
done.